pkg/operator/ceph/cluster/cluster.go

```diff
@@ -134,70 +142,87 @@ func (c *cluster) detectCephMajorVersion(image string, timeout time.Duration) (s
 
 func (c *cluster) createInstance(rookImage string) error {
 
-	// Create a configmap for overriding ceph config settings
-	// These settings should only be modified by a user after they are initialized
-	placeholderConfig := map[string]string{
-		k8sutil.ConfigOverrideVal: "",
-	}
-	cm := &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: k8sutil.ConfigOverrideName,
-		},
-		Data: placeholderConfig,
-	}
-	k8sutil.SetOwnerRef(c.context.Clientset, c.Namespace, &cm.ObjectMeta, &c.ownerRef)
+	defer c.setOrchestrationRunning(false)
 
-	_, err := c.context.Clientset.CoreV1().ConfigMaps(c.Namespace).Create(cm)
-	if err != nil && !errors.IsAlreadyExists(err) {
-		return fmt.Errorf("failed to create override configmap %s. %+v", c.Namespace, err)
+	if c.getOrchestrationRunning() == true {
+		logger.Debugf("As createInstance is currently running added this request as pending.")
+		c.setOrchestrationPending(true)
+		return nil
 	}
 
-	// Start the mon pods
-	c.mons = mon.New(c.Info, c.context, c.Namespace,
-		c.Spec.DataDirHostPath, rookImage, c.Spec.CephVersion, c.Spec.Mon,
-		cephv1.GetMonPlacement(c.Spec.Placement), c.Spec.Network.HostNetwork,
-		cephv1.GetMonResources(c.Spec.Resources), c.ownerRef)
-	clusterInfo, err := c.mons.Start()
-	if err != nil {
-		return fmt.Errorf("failed to start the mons. %+v", err)
-	}
-	c.Info = clusterInfo // mons return the cluster's info
+	for c.getOrchestrationPending() == true || c.getOrchestrationRunning() == false {
+		c.setOrchestrationRunning(true)
```

Comment:
I think the `c.getOrchestrationPending() == true || c.getOrchestrationRunning() == false` and `c.setOrchestrationRunning(true)` steps need to be an atomic set-and-check. Otherwise, two simultaneous threads may both see a truthful condition close to the same time before one thread can lock `c.setOrchestrationRunning(true)`. Since the `get` methods are implemented as a read lock, any number of threads can read the value at once.

Eliminating the for loop and instead just using a check-and-set with a read-write lock will be sufficient and eliminate some cyclomatic complexity and small processing overhead present in the for loop. Each thread will block on the atomic check-and-set until they are able to obtain the lock, and the scheduler will handle the checking and rechecking of the values.

Looking further, I think the checks and *both* sets (including `c.setOrchestrationPending(false)`) need to be an atomic operation.

Reply:
In did a quick measurement regarding the run-time of `createInstance()`. In my cluster it took about 2.5 minutes (3 worker nodes + 1 master node). Blocking the threads would mean to accumulate this execution time over the waiting threads.
@travisn Any thoughts? Get rid of the loop or try to eliminate the possible race-conditions?

Reply:
The key is that we only want one orchestration running at a time. If two orchestrations are able to start, they are likely to cause issues.

@shaas Is your concern about the second request blocking for a long period of time while the first orchestration is completing? I don't think that's an issue if we are careful. If we get/set the "orchestrationPending" safely, then the loop will pick up the second orchestration as soon as the first one is done. And if a third or fourth orchestration request comes before the first one is completed, only one more orchestration will be run with requests 2-4 combined.

Reply:
One more thought is that we would need to make sure that `c.Spec` is not updated by the new request until the first orchestration is done. The spec should likely be passed as a param to `createInstance()`, then the locking code around `orchestrationPending` would need to decide when to set `c.Spec`.

Reply:
> The key is that we only want one orchestration running at a time. If two orchestrations are able to start, they are likely to cause issues.
> 
> @shaas Is your concern about the second request blocking for a long period of time while the first orchestration is completing? I don't think that's an issue if we are careful. If we get/set the "orchestrationPending" safely, then the loop will pick up the second orchestration as soon as the first one is done. And if a third or fourth orchestration request comes before the first one is completed, only one more orchestration will be run with requests 2-4 combined.

Exactly! That is what I was trying to implement.

Reply:
> One more thought is that we would need to make sure that `c.Spec` is not updated by the new request until the first orchestration is done. The spec should likely be passed as a param to `createInstance()`, then the locking code around `orchestrationPending` would need to decide when to set `c.Spec`.

Good Point! I will change that accordingly.